### ISSUE TITLE
More blocks addon

### ISSUE THREAD

**Issue by wgyt735yt** [ScratchAddons/ScratchAddons#817]
**Describe the new addon**

This addon adds custom blocks and hacked blocks to the Scratch editors under the Scratch addons category.
This addon could also allow other addons like the pause addon to add their blocks (sa-pause in this example) to this category.
![image](https://user-images.githubusercontent.com/68466727/99908012-0123ff00-2cae-11eb-9525-ee345f72f9ba.png)

**Comment by apple502j** [comment]
I'd say it's rejected - we don't want to encourage Scratchers to install Scratch Addons to run some projects. Rules have changed since NitroBlock.

**Comment by wgyt735yt** [comment]
> I'd say it's rejected - we don't want to encourage Scratchers to install Scratch Addons to run some projects. Rules have changed since NitroBlock.

No, I don't think you understand what I'm saying.
The pause addon takes advantage of a custom block called sa-pause, and that could go in this category. This would also make it easier to use common workarounds for commonly suggested blocks and make them easy to add.

**Comment by apple502j** [comment]
Pause block is a custom block, and does not require new extension category, and it's only for debugging purposes.

**Comment by wgyt735yt** [comment]
It only uses vanilla blocks, and it only makes it easy to drag in workarounds like the ones listed here https://scratch.mit.edu/discuss/topic/444061/

**Comment by JeffaloBob** [comment]
> Pause block is a custom block, and does not require new extension category, and it's only for debugging purposes.

debugging purposes for project creators though.

i think this is a good idea

**Comment by wgyt735yt** [comment]
Here is what needs to happen for this to become reality:

- [ ] We need to inject a new category to the sidebar

- [ ] We need a way to add blocks to the sidebar

- [ ] We need a way for other addons to add their own blocks

If anyone has suggestions let me know

**Comment by wgyt735yt** [comment]
Progress on the prototyping side has been made :)
I made the category, the block design and I made a video demonstrating.
Due to GitHub not allowing me to upload the mp4 file, I have the video hosted on Google Drive and it is available [here](https://drive.google.com/file/d/1X2dDcryFL2PbYRx_r1iHSIxliXmXGKTW/view).

**Comment by Explosion-Scratch** [comment]
Just the stuff like while loops, for i in range, etc would be really helpful. No support however for blocks that need scratchaddons to be run, that would attract ST attention and be really bad. Speaking of which, should the ```SA-pause``` be removed?

**Comment by TheColaber** [comment]
> Speaking of which, should the `SA-pause` be removed?

No! Its purpose was to debug and let me tell you, if I had this feature when I first started, I would have probably learned a lot more, then to be stuck on tons of bugs.

**Comment by Explosion-Scratch** [comment]
Alright, here is my list:

- [ ] While obsolete block
- [ ] For i in range obsolete block
- [ ] Counter obsolete blocks
- [ ] Stop sound hacked block (Credit to qucchia)
- [ ] Point towards block, there is a one line method.
- [ ] White blob block, just for fun.

I'll make a project with these, just a sec.

**Comment by apple502j** [comment]
@Explosion-Scratch sa-pause will NOT be removed, but that's just because it's only useful when debugging and we expect people to remove it before sharing.

What's the "hacked block"?

Also, no fun/no-op blocks, please.

**Comment by Explosion-Scratch** [comment]
Got it! Okay!

**Comment by Explosion-Scratch** [comment]
Here's the project that I made, you'll have to use forkphorus downloader. https://scratch.mit.edu/projects/454744006/editor

**Comment by apple502j** [comment]
@Explosion-Scratch  Generally speaking, Scratch Addons should not "corrupt" a project i.e. make it unusable for those w/o it. Seems like editing the "sound hacked block" in vanilla causes soft crash, so that's a no. Editing Scratch 2.0 hacked block did not cause an error, AFAIK. Not sure whether I should allow "custom block definition with input", though.

No-op blocks are also no; including red hat blocks and "all at once" (this is just an if block that always executes. if you want to mass-delete clone, broadcast is the way. did you really test the behavior?)

"Secret masking" is no, as it will probably catch ST's eyes and cause trouble, like shadow blocks did.

**Comment by mxmou** [comment]
> ... cause trouble, like shadow blocks did.

They did?

**Comment by WorldLanguages** [comment]
Well, shadow blocks don't corrupt a project but it's a terrible idea to allow users to easily hide blocks.

**Comment by Explosion-Scratch** [comment]
> > ... cause trouble, like shadow blocks did.
> 
> They did?

Yup. I made one of my platformers have 'no' blocks because I hid them with shadow. Even though the actual code for the project was available in another project the ST still gave me an alert and took it down.

**Comment by Explosion-Scratch** [comment]
> Well, shadow blocks don't corrupt a project but it's a terrible idea to allow users to easily hide blocks.

I guess so.

**Comment by Explosion-Scratch** [comment]
> @Explosion-Scratch Generally speaking, Scratch Addons should not "corrupt" a project i.e. make it unusable for those w/o it. Seems like editing the "sound hacked block" in vanilla causes soft crash, so that's a no. Editing Scratch 2.0 hacked block did not cause an error, AFAIK. Not sure whether I should allow "custom block definition with input", though.
> 

True, what about the ones that are actually useful though? Like the 'for i in range' one, those are super useful. Also about the 'custom block with input' I think that this could actually be very useful, because I have had scripts that I worked hard on stolen with no credit before.

> No-op blocks are also no; including red hat blocks 

They're cool, but I understand.

> and "all at once" (this is just an if block that always executes. if you want to mass-delete clone, broadcast is the way. did you really test the behavior?)

No. I just did that based on the comment on the original project I used it from.

> "Secret masking" is no, as it will probably catch ST's eyes and cause trouble, like shadow blocks did.

That makes sense.

**Comment by MasimbaTheHuman** [comment]
We need to make sure that all of these blocks only make functions more convenient, instead of adding new features.

**Comment by wgyt** [comment]
> Just the stuff like while loops, for I in range, etc would be really helpful. No support however for blocks that need scratchaddons to be run, that would attract ST attention and be really bad. Speaking of which, should the `SA-pause` be removed?

This doesn't need SA, it would add the blocks to the project, these would be vanilla blocks.


> @Explosion-Scratch Generally speaking, Scratch Addons should not "corrupt" a project i.e. make it unusable for those w/o it. Seems like editing the "sound hacked block" in vanilla causes a soft crash, so that's a no. Editing Scratch 2.0 hacked block did not cause an error, AFAIK. Not sure whether I should allow "custom block definition with input", though.
> 
> No-op blocks are also no; including red hat blocks and "all at once" (this is just an if block that always executes. if you want to mass-delete clone, a broadcast is the way. did you really test the behavior?)
> 
> "Secret masking" is no, as it will probably catch ST's eyes and cause trouble as shadow blocks did.

This again would only add and use vanilla blocks, except sa-pause, which doesn't work w/o SA


> Well, shadow blocks don't corrupt a project but it's a terrible idea to allow users to easily hide blocks.

what? hide blocks? ok, that isn't coming anytime soon :P

And on another note:
all at once and other useful (but obsolete) blocks are hidden in the code of scratch blocks

**Comment by Marc92020** [comment]
This would be cool however the only problem would be that other scratchers would not be able to see these blocks and the red hat block can't be connected to any block so the only workaround would be to only have them for unshared projects. Or just use hacked blocks that already exist like the "while <>" and the "stop sound ()" block.

**Comment by wgyt** [comment]
> This would be cool however the only problem would be that other scratchers would not be able to see these blocks and the red hat block can't be connected to any block so the only workaround would be to only have them for unshared projects. Or just use hacked blocks that already exist like the "while <>" and the "stop sound ()" block.

Other Scratchers WOULD be able to see all of the blocks because for the five millionth time, this only uses vanilla blocks that scratch blocks can handle natively like while <>, all at once, and custom blocks that Scratch Addons would add, that only use native blocks.

**Comment by Marc92020** [comment]
> > This would be cool however the only problem would be that other scratchers would not be able to see these blocks and the red hat block can't be connected to any block so the only workaround would be to only have them for unshared projects. Or just use hacked blocks that already exist like the "while <>" and the "stop sound ()" block.
> 
> Other Scratchers WOULD be able to see all of the blocks because for the five millionth time, this only uses vanilla blocks that scratch blocks can handle natively like while <>, all at once, and custom blocks that Scratch Addons would add, that only use native blocks.

ok so other scratchers who don't have the extension could only see the workarounds?

**Comment by wgyt** [comment]
> > > This would be cool however the only problem would be that other scratchers would not be able to see these blocks and the red hat block can't be connected to any block so the only workaround would be to only have them for unshared projects. Or just use hacked blocks that already exist like the "while <>" and the "stop sound ()" block.
> > 
> > 
> > Other Scratchers WOULD be able to see all of the blocks because for the five millionth time, this only uses vanilla blocks that scratch blocks can handle natively like while <>, all at once, and custom blocks that Scratch Addons would add, that only use native blocks.
> 
> ok so other scratchers who don't have the extension could only see the workarounds?

they would see all the blocks and they would work, except sa-pause
@Marc92020

**Comment by lisa-wolfgang** [comment]
Let me try to explain what I think @wgyt means...

In the blocks palette, there would be a Scratch Addons category. It would look something like this:

![image](https://user-images.githubusercontent.com/43426138/106219843-d9ff1780-619f-11eb-9810-0bf7ecd1e954.png)

But when the user drags the block into the editor, it's replaced with an entire stack of code, like this:

![image](https://user-images.githubusercontent.com/43426138/106220431-1c752400-61a1-11eb-8a74-19bbb90aa46f.png)

The idea is to provide frequently used sequences of code in the block palette for convenient access. Sequences such as the workaround for the "for i in range" block (NOT the actual block! The workaround sequence!) would obviously go here instead of my terrible example.

It would be really cool (an update-defining feature) if the user was allowed to add their OWN sequences to the Scratch Addons category, for usage across all sprites and projects.

**Comment by wgyt** [comment]
> Let me try to explain what I think @wgyt means...
> 

This is what I meant. This is what I was trying to explain, I might try to make a prototype sometime soon.

**Comment by RedGuy12** [comment]
The `sa-pause` block from the pause button addon can be there too
________________________________

**Comment by CuriousG2** [comment]
If the block required inputs, it could also turn into a custom block.

**Comment by sfederici** [comment]
Thanks to @RedGuy12 to suggest to have a look at this post. I was thinking of a similar "vanilla blocks" addon, but I'd like

1. to add vanilla blocks (e.g. the FOREVER IF block that since Scratch 2.0 was removed from the palette) in the standard Scratch categories
2. the vanilla block to show in the script area exactly as they show in the palette (keep reading :-)

They won't be a problem for non-ScratchAddons users, as they will be saved as their corresponding Scratch code (as in https://github.com/ScratchAddons/ScratchAddons/issues/817#issuecomment-769518560) plus a comment that will allow to rebuild the vanilla block once the sb3 project is loaded in the editor if the "vanilla blocks addon" is enabled.

**Comment by Sly-Little-Fox** [comment]
I think that it should not create a separate category. Maybe make it put these obsolete blocks where they were in the past?

**Comment by wgyt** [comment]
> I think that it should not create a separate category. Maybe make it put these obsolete blocks where they were in the past?

Possibly? But workarounds go in a different category

**Comment by Sly-Little-Fox** [comment]
> > I think that it should not create a separate category. Maybe make it put these obsolete blocks where they were in the past?
> 
> Possibly? But workarounds go in a different category

Probably yes. Another solution is to add an icon that this block is added my SA and is a workaround.

**Comment by lisa-wolfgang** [comment]
I'm going to start a running list of ideas for workarounds to include in this category -- just a brainstormed list from a variety of sources and not filtered (except for already rejected ones or workarounds that are too large in scope for just one script):

- Point towards x: __ y: __
- Bounce (not just edge, but anywhere on the stage)
- Switch to previous costume (unused block in vanilla available)
- Width / height of current costume
- Next / previous sound
- Pause / resume sound
- When stop sign clicked
- When costume switches to ____
- While ____ (unused block in vanilla available)
- If / else if
- counter / increment counter / reset counter (unused blocks in vanilla available)
- For i in range (unused block in vanilla available)
- For each (iterate over lists)
- Delete clone (with notch on the bottom)
- "Is clone?" boolean
- "Turbo Mode enabled?" boolean
- "Frames per second" reporter
- XOR operator
- "True" boolean
- "False" boolean
- Letters __ to __ of ____
- ≤ / ≥ operators
- sign of __ (1 for positive, 0 for zero, -1 for negative)
- item [last] of [list] (unused block in vanilla available)
- copy list [ ] to list [ ]
- item [random] of [list]
- // ____ (comment block)

**Comment by apple502j** [comment]
Rejected:

- Bounce
- Width / height of current costume
- Next/previous sound
- Pause / resume sound
- When stop sign clicked
- When costume switches to ____
- "Is clone?" boolean
- "Turbo Mode enabled?" boolean
- "Frames per second" reporter
- XOR operator
- "True" boolean
- Letters __ to __ of ____

**Comment by apple502j** [comment]
My opinion on this:

- Addons that add existing but hidden blocks should be fine.
- Addons that add a custom block for everyone should also be fine.
- Addons that transform existing blocks to look like shorter SA blocks? Hmmm.
- Addons that transform existing script to look like shorter SA blocks? No. Just use vanilla custom blocks. (This means "custom reporters" must be implemented without any stack blocks.)
- Also, no custom hat blocks for technical reasons.
- No blocks that require Scratch Addons to work reasonably fast and accurately. This includes costume size and "Turbo mode enabled" block.

**Comment by ego-lay-atman-bay** [comment]
> * item [last] of [list] (unused block in vanilla available)
> * item [random] of [list]

those aren't really unused since it's super easy to implement yourself. All you have to do is paste the text into the input, or just put a join reporter in the input and type in the text. I've actually used it many times.
![block_9_21_2021-6_43_48 PM|479x500, 75%](https://user-images.githubusercontent.com/70973578/134265770-8b879946-57e6-4143-be59-8c48f9e46e50.png)

**Comment by Cool-Scratcher** [comment]
> those aren't really unused since it's super easy to implement yourself. All you have to do is paste the text into the input, or just put a join reporter in the input and type in the text. I've actually used it many times.
You could also do that to get more keys in the "if key pressed" blocks, but that still became an addon because it was simply easier to have it in the block. I think a similar thing applies here.

**Comment by ego-lay-atman-bay** [comment]
> > those aren't really unused since it's super easy to implement yourself. All you have to do is paste the text into the input, or just put a join reporter in the input and type in the text. I've actually used it many times.
> 
> You could also do that to get more keys in the "if key pressed" blocks, but that still became an addon because it was simply easier to have it in the block. I think a similar thing applies here.

pretty sure the addon still doesn't add all the keys you can use though.

**Comment by RedGuy12** [comment]
Speaking of that, maybe this addon should be merged with the more keys addon...🤔


________________________________

**Comment by RedGuy12** [comment]
It adds all that are available on most keyboards


________________________________

**Comment by Tacodude7729** [comment]
@RedGuy12 
> Speaking of that, maybe this addon should be merged with the more keys addon...thinking

I'm not too sure how you would do that, given this is a text input not a dropdown. We could make the plugin more general, something like 'reveal hidden block options' but unless there where a few more things to go in it I think it makes more scene to just have them as two plugins. 

@Cool-Scratcher 

> You could also do that to get more keys in the "if key pressed" blocks, but that still became an addon because it was simply easier to have it in the block. I think a similar thing applies here.

I'm pretty sure it's not possible without modding the `project.json` to get the extra keys in the `when () key pressed` event block.

@ego-lay-atman-bay 

> pretty sure the addon still doesn't add all the keys you can use though.

See #3480

**Comment by sfederici** [comment]
> I'm not too sure how you would do that, given this is a text input not a dropdown. We could make the plugin more general, something like 'reveal hidden block options' but unless there where a few more things to go in it I think it makes more scene to just have them as two plugins.

I must say that I really hate the way some of the blocks in 3.0 lost their dropdowns. If they can be added back, that would be really great.

**Comment by wgyt** [comment]
> Speaking of that, maybe this addon should be merged with the more keys addon...🤔
> […](#)

More keys was one of the original goals of more blocks iirc, but if this would be merged, it would need other stuff in it other than more keys like workaround blocks

**Comment by Cool-Scratcher** [comment]
> > You could also do that to get more keys in the "if key pressed" blocks, but that still became an addon because it was simply easier to have it in the block. I think a similar thing applies here.
> 
> I'm pretty sure it's not possible without modding the `project.json` to get the extra keys in the `when () key pressed` event block.

It's definitely possible without modifying anything, I use it all the time. Put a join block in the key pressed and you can then detect those other keys by typing the key name in one of the options of the  join block, just like the list blocks

**Comment by return-misaPuding** [comment]
> Well, shadow blocks don't corrupt a project but it's a terrible idea to allow users to easily hide blocks.

Depends on whether non-SA users also won't see the content of these blocks. If they won't, the project would be illegal because it won't fulfill the requirements of the CC-BY-SA licence (because you won't be able to see the full content of the project as the content of that hacked reporter block won't be visible), so ScratchAddons would essentialy allow the user to make their project illegal

**Comment by WorldLanguages** [comment]
Yeah, we definitely don't want users to easily hide blocks

**Comment by hueychen27** [comment]
Here's a suggestion: Drag the block and a custom block with some blocks will appear at your cursor. That way you other non-SA users can see the code. If the custom block name is already mentioned, then add like a "Bounce2" or something like that.

**Comment by DNin01** [comment]
> My opinion on this:
> 
> - Addons that add existing but hidden blocks should be fine.
> 
> - Addons that add a custom block for everyone should also be fine.
> 
> (...)

I agree with everything here but I personally wouldn't suggest revealing the "lost" blocks such as the while loop or the counter.

**Comment by BroJac5246** [comment]
> > My opinion on this:
> > 
> > - Addons that add existing but hidden blocks should be fine.
> > - Addons that add a custom block for everyone should also be fine.
> > 
> > (...)
> 
> I agree with everything here except I wouldn't reveal the "lost" blocks such as the while loop or the counter - that could look like we're encouraging Scratchers to violate section 4.4 on Scratch's Terms of Use.

The extension as a whole already does this, so is that a huge problem? And the blocks are native Scratch blocks. If ST wanted, they could display a message that states "This project was made for an outdated version of Scratch and is unsupported." Also, ST doesn't even remove projects on Scratch that advertise hacked blocks, so they're not really doing anything about it.

That said, this obviously isn't legal advice and is just my impression.

**Comment by WorldLanguages** [comment]
> 4.4 You may only submit user-generated projects that were created with (1) the Scratch website editor or (2) an unmodified copy of the Scratch editor compiled from the source code described in Section 5.3.

> 5.3 The source code for Scratch 1.4 is available for download and subject to the copyright notice as indicated on the [Scratch FAQ](https://scratch.mit.edu/faq) page.

So according to Scratch's Terms of Service, you may not submit projects that were created with Scratch Dekstop, the official offline editor for 3.0 ¯\\_(ツ)_/¯

**Comment by sfederici** [comment]
They might have to rewrite it

**Comment by Secret-chest** [comment]
Or even the unofficial Scratch Flatpak which generates identical projects
